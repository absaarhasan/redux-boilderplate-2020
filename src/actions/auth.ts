import * as actions from 'Actions/constants';
import {
  LoginAction,
  LogoutAction,
  RegisterAction,
  ForgotPasswordAction,
  VerifyPasswordAction,
  IsAuthorisedAction,
  SetUserSessionAction,
  LoginData,
  RegData,
  ForgotData,
  VerifyData,
} from './types';

export const loginAction = (data: LoginData): LoginAction => ({
  type: actions.AUTH_LOGIN,
  data,
});

export const logoutAction = (): LogoutAction => ({
  type: actions.AUTH_LOGOUT,
});

export const registerAction = (data: RegData): RegisterAction => ({
  type: actions.AUTH_REGISTER,
  data,
});

export const forgotPasswordAction = (data: ForgotData): ForgotPasswordAction => ({
  type: actions.AUTH_FORGOT_PASSWORD,
  data,
});

export const verifyPasswordAction = (data: VerifyData): VerifyPasswordAction => ({
  type: actions.AUTH_VERIFY_PASSWORD,
  data,
});

export const isAuthorisedAction = (): IsAuthorisedAction => ({
  type: actions.AUTH_IS_AUTHORISED,
});

export const setUserSessionAction = (data: {}): SetUserSessionAction => ({
  type: actions.USER_SESSION_SET,
  data,
});

export const clearUserSessionAction = (): SetUserSessionAction => ({
  type: actions.USER_SESSION_CLEAR,
});
