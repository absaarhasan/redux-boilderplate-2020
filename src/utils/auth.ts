import awsConfig from 'Constants/awsConfig';
import * as AmazonCognitoIdentity from 'amazon-cognito-identity-js';
import {
  RegData, ForgotData, VerifyData, LoginData,
} from 'Actions/types';

const poolData = {
  UserPoolId: awsConfig.cognito.userPoolId,
  ClientId: awsConfig.cognito.clientId,
};

const userPool = new AmazonCognitoIdentity.CognitoUserPool(poolData);

export type VerifyPassword = ReturnType<typeof verifyPasswordService>

export const verifyPasswordService = (data: VerifyData) => {
  const { email, password, code } = data;
  const userData = {
    Username: email,
    Pool: userPool,
  };

  const cognitoUser = new AmazonCognitoIdentity.CognitoUser(userData);

  return new Promise((resolve, reject) => {
    cognitoUser.confirmPassword(code, password, {
      onSuccess: () => {
        resolve();
      },
      onFailure: (err) => {
        reject(err);
      },
    });
  }).catch((err) => {
    throw new Error(err.message || JSON.stringify(err));
  });
};

export const forgotPasswordService = ({ email }: ForgotData) => {
  const userData = {
    Username: email,
    Pool: userPool,
  };

  const cognitoUser = new AmazonCognitoIdentity.CognitoUser(userData);

  return new Promise((resolve, reject) => {
    cognitoUser.forgotPassword({
      onSuccess: (result) => {
        resolve(result);
      },
      onFailure: (err) => {
        reject(err);
      },
    });
  }).catch((err) => {
    throw new Error(err.message || JSON.stringify(err));
  });
};

export const logoutService = () => {
  const cognitoUser = userPool.getCurrentUser();
  if (cognitoUser) {
    cognitoUser.signOut();
  }
};

export const isAuthorisedService = () => {
  const cognitoUser = userPool.getCurrentUser();
  if (cognitoUser) {
    return new Promise((resolve, reject) => {
      cognitoUser.getSession((err, session) => {
        if (err) {
          reject(err);
        }
        cognitoUser.getUserAttributes((err, result) => {
          if (err) {
            reject(err);
          }
          const authData = {};
          result.forEach((result) => {
            authData[result.getName()] = result.getValue();
          });
          resolve(authData);
        });
      });
    }).catch((err) => {
      throw new Error(err.message || JSON.stringify(err));
    });
  }
  return null;
};

export const loginService = (data: LoginData) => {
  const authenticationDetails = new AmazonCognitoIdentity.AuthenticationDetails(
    data,
  );
  const userData = {
    Username: data.Username,
    Pool: userPool,
  };
  const cognitoUser = new AmazonCognitoIdentity.CognitoUser(userData);

  return new Promise((resolve, reject) => {
    cognitoUser.authenticateUser(authenticationDetails, {
      onSuccess: (result) => {
        const accessToken = result.getAccessToken().getJwtToken();
        resolve(accessToken);
      },
      onFailure: (err) => {
        reject(err);
      },
    });
  }).catch((err) => {
    throw new Error(err.message || JSON.stringify(err));
  });
};

export const registerService = (data: RegData) => {
  const {
    userNameObj, emailObj, email, password,
  } = data;
  const attrEmail = new AmazonCognitoIdentity.CognitoUserAttribute(emailObj);
  const attrUserName = new AmazonCognitoIdentity.CognitoUserAttribute(
    userNameObj,
  );
  const attributeList = [attrEmail, attrUserName];

  return new Promise((resolve, reject) => {
    userPool.signUp(email, password, attributeList, null, (err, result) => {
      result ? resolve(result) : reject(err);
    });
  }).catch((err) => {
    throw new Error(err.message || JSON.stringify(err));
  });
};
