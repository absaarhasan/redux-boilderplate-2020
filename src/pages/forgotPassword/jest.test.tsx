import React from 'react';
import { cleanup, screen, fireEvent } from '@testing-library/react';
import * as ReactRedux from 'react-redux';
import * as actions from 'Actions/auth';
import { render } from '../../../jest.setup';
import ForgotPassword from '.';

beforeEach(() => {
  // render(<ForgotPassword />);
});

afterEach(cleanup);

describe('<ForgotPassword /> spec', () => {
  it('should render', () => {
    render(<ForgotPassword />);
    expect(screen.getByTestId('forgot-container')).toMatchSnapshot();
  });
  it('It should display the required form fields', () => {
    render(<ForgotPassword />);
    expect(screen.getByTestId('email-field')).toBeDefined();
  });
  it('the button should submit the form', () => {
    const useDispatch = jest.spyOn(ReactRedux, 'useDispatch');
    const actionCreator = jest.spyOn(actions, 'forgotPasswordAction');
    render(<ForgotPassword />);
    fireEvent.click(screen.getByTestId('submit-button'));
    expect(useDispatch).toHaveBeenCalled();
    expect(actionCreator).toHaveBeenCalled();
  });
});
