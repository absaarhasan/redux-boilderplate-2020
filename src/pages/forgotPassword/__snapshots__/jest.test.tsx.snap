// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ForgotPassword /> spec should render 1`] = `
.emotion-3 {
  display: grid;
  grid-template-columns: 90%;
  grid-template-rows: 8em;
  grid-auto-rows: 3em;
  gap: 1em 1%;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  justify-items: start;
}

@media (min-width:600px) {
  .emotion-3 {
    grid-template-columns: 25% 25%;
  }
}

@media (min-width:920px) {
  .emotion-3 {
    grid-template-columns: 25% 25% 25%;
  }
}

.emotion-0 {
  grid-column: 1;
  grid-row: 0 / 2;
  justify-self: center;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  font-size: 2em;
  color: #04080c;
}

@media (min-width:600px) {
  .emotion-0 {
    grid-column: 1 / span 2;
  }
}

@media (min-width:920px) {
  .emotion-0 {
    grid-column: 1 / span 3;
  }
}

.emotion-1 {
  justify-self: stretch;
  -webkit-align-self: stretch;
  -ms-flex-item-align: stretch;
  align-self: stretch;
  font-size: 1em;
  padding: 0 1%;
  color: #778189;
}

.emotion-2 {
  justify-self: stretch;
  -webkit-align-self: stretch;
  -ms-flex-item-align: stretch;
  align-self: stretch;
  font-size: 1em;
  color: #778189;
  grid-column: 1;
}

@media (min-width:600px) {
  .emotion-2 {
    justify-self: stretch;
    grid-column: 1 / span 2;
  }
}

@media (min-width:920px) {
  .emotion-2 {
    justify-self: stretch;
    grid-column: auto;
  }
}

<div
  class="emotion-3"
  data-testid="forgot-container"
>
  <h1
    class="emotion-0"
  >
    Forgot Password
  </h1>
  <input
    class="emotion-1"
    data-testid="email-field"
    id="email"
    name="email"
    placeholder="Email address"
    required=""
    type="text"
    value=""
  />
  <button
    class="emotion-2"
    data-testid="submit-button"
    type="button"
  >
    Send Verification Code
  </button>
</div>
`;
